src/Features/Core/Portable/FindReferences/IDefinitionsAndReferencesFactory.cs

```diff
@@ -0,0 +1,273 @@
+                            if (documentLocation.CanNavigateTo())
+                            {
+                                result.Add(DefinitionLocation.CreateDocumentLocation(documentLocation));
+                            }
+                        }
+                    }
+                }
+            }
+
+            if (result.Count == 0)
+            {
+                // If we got no definition locations, then create a sentinel one
+                // that we can display but which will not allow navigation.
+                result.Add(DefinitionLocation.CreateNonNavigatingLocation(
+                    DefinitionLocation.GetOriginationParts(definition)));
+            }
+
+            return result.ToImmutable();
+        }
+
+        private static ImmutableArray<Location> FilterDefinitionLocations(ISymbol definition)
+        {
+            // When finding references of a namespace, the data provided by the ReferenceFinder
+            // will include one definition location for each of its exact namespace
+            // declarations and each declaration of its children namespaces that mention
+            // its name (e.g. definitions of A.B will include "namespace A.B.C"). The list of
+            // reference locations includes both these namespace declarations and their
+            // references in usings or fully qualified names. Instead of showing many top-level
+            // declaration nodes (one of which will contain the full list of references
+            // including declarations, the rest of which will say "0 references" due to
+            // reference deduplication and there being no meaningful way to partition them),
+            // we pick a single declaration to use as the top-level definition and nest all of
+            // the declarations & references underneath.
+            if (definition.IsKind(SymbolKind.Namespace))
+            {
+                // Prefer source location over metadata.
+                var firstLocation = definition.Locations.FirstOrDefault(loc => loc.IsInSource) ?? definition.Locations.First();
+                return ImmutableArray.Create(firstLocation);
+            }
+
+            return definition.Locations;
+        }
+
+        private static void CreateReferences(
+            ReferencedSymbol referencedSymbol,
+            ImmutableArray<SourceReferenceItem>.Builder references,
+            DefinitionItem definitionItem,
+            HashSet<ValueTuple<Document, TextSpan>> uniqueLocations)
+        {
+            foreach (var referenceLocation in referencedSymbol.Locations)
+            {
+                var location = referenceLocation.Location;
+                Debug.Assert(location.IsInSource);
+
+                var document = referenceLocation.Document;
+                var sourceSpan = location.SourceSpan;
+
+                if (uniqueLocations.Add(ValueTuple.Create(document, sourceSpan)))
```

Comment:
Why not just create the DocumentLocation directly and use it in `uniqueLocations`?